dev/data/fineweb.py

```diff
@@ -99,7 +120,7 @@ def tokenize(doc):
             remainder = args.shard_size - token_count
             progress_bar.update(remainder)
             all_tokens_np[token_count:token_count+remainder] = tokens[:remainder]
-            write_datafile(filename, all_tokens_np)
+            write_datafile(filename, list(all_tokens_np), args.model)
```

Comment:
why convert to list

Reply:
i simplified `write_datafile` so that it doesn't have to handle both numpy & lists, it's cleaner i think

Reply:
There could be many tokens, so creating a Python list could be very wasteful

Reply:
```python
>>> a = np.random.randn(10)
>>> a
array([-1.39200423,  0.91909499,  0.49247546,  0.73578011, -0.46485352,
        0.06844696,  1.21521025,  0.18951044, -0.33376094,  1.03115886])
>>> list(a)
[-1.3920042324598616, 0.9190949922347375, 0.49247545796208686, 0.7357801064341112, -0.4648535191489631, 0.06844695804812885, 1.2152102515229188, 0.18951044050354424, -0.33376094056177236, 1.0311588596558752]
>>> z = list(a)
>>> z[0]
-1.3920042324598616
>>> type(z[0])
<class 'numpy.float64'>
```

Reply:
vs `tolist`

```python
>>> a = np.random.randn(10)
>>> a
array([-0.28416783,  3.61778557,  0.45557321,  0.6585392 , -0.54974637,
       -0.50662981,  0.36080734,  0.76378507, -1.60443242,  0.41719901])
>>> a.tolist()
[-0.2841678282966355, 3.6177855666263548, 0.45557321422210056, 0.6585391952854299, -0.5497463693208792, -0.5066298099246493, 0.36080734397795633, 0.7637850737170351, -1.6044324246329673, 0.4171990143035489]
>>> z = a.tolist()
>>> z[0]
-0.2841678282966355
>>> type(z[0])
<class 'float'>
```